components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt

```diff
@@ -26,16 +27,93 @@ internal object TabListReducer {
                 )
             }
 
-            is TabListAction.SelectTabAction -> state.copy(selectedTabId = action.tabId)
+            is TabListAction.SelectTabAction -> {
+                state.copy(selectedTabId = action.tabId)
+            }
 
             is TabListAction.RemoveTabAction -> {
                 val tab = state.findTab(action.tabId)
+
                 if (tab == null) {
                     state
                 } else {
-                    state.copy(tabs = state.tabs - tab)
+                    val previousIndex = state.tabs.indexOf(tab)
+                    val updatedTabList = state.tabs - tab
+                    val updatedSelection = if (state.selectedTabId == tab.id) {
+                        findNewSelectedTabId(updatedTabList, tab.content.private, previousIndex)
+                    } else {
+                        state.selectedTabId
+                    }
+
+                    state.copy(
+                        tabs = updatedTabList,
+                        selectedTabId = updatedSelection
+                    )
                 }
             }
         }
     }
 }
+
+/**
+ * Find a new selected tab and return its id after the tab at [previousIndex] was removed.
+ */
+private fun findNewSelectedTabId(
+    tabs: List<TabSessionState>,
+    isPrivate: Boolean,
+    previousIndex: Int
+): String? {
+    if (tabs.isEmpty()) {
+        // There's no tab left to select.
+        return null
+    }
+
+    val predicate: (TabSessionState) -> Boolean = { tab -> tab.content.private == isPrivate }
```

Comment:
Can you omit the type if you do this?
```suggestion
    val predicate = { tab: TabSessionState -> tab.content.private == isPrivate }
```

Reply:
Yeah, that works. But do you think it's more readable? If I omit the type then Android Studio will show it itself and there's actually more text on the screen after the change :D

Reply:
I didn't realize Android Studio does that, feel free to keep it as you originally wrote it.